aten/src/THC/THCDeviceUtils.cuh

```diff
@@ -107,6 +107,15 @@ __device__ __forceinline__ T WARP_SHFL_DOWN(T value, unsigned int delta, int wid
     return __shfl_down(value, delta, width);
 #endif
 }
+template <>
+__device__ __forceinline__ at::Half WARP_SHFL_DOWN(at::Half value, unsigned int delta, int width, unsigned int mask)
```

Comment:
It looks like WARP_SHFL_DOWN should never be called with Half argument, so why is this necessary?

Reply:
It's called here in Reduce.cuh:
https://github.com/pytorch/pytorch/pull/11908/files#diff-662693ef7b7f32fa32d7179b6614fc16R296

(The Reduce.cuh file might be collapsed by default in GitHub's "Files changed" view)

Reply:
But `warp_reduce` is called on the argument of type `arg_t` and `arg_t` is https://github.com/pytorch/pytorch/pull/11908/files#diff-662693ef7b7f32fa32d7179b6614fc16R199 (should be no narrower than fp32). I'm (we're) very suspect of any operations directly on halfs, it's precision/overflow/underflow risk. With current reduction ops we made sure that it never happens (all data is converted on read), and it would be great if it did not creep back in.

Reply:
I'll change it to use float32 for intermediate values, but I don't really see how that's going to help. The output is still float16 so you still have overflow issues. You can see this in master with:

```
x = torch.ones(65520, device='cuda', dtype=torch.float16)
x.sum()
```

```
tensor(inf, device='cuda:0', dtype=torch.float16)
```

Reply:
Nevermind. You're right, it already uses float32 for float16 inputs....